src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -292,8 +311,81 @@ internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort a
                     _ => hex
                 };
 
-                return new MappedField(name, this, type, key.Data);
-            });
+                return new MappedField(name, @this, type, data);
+            },
+            this);
+        }
+
+        /// <summary>
+        /// Gets the field of <see cref="DataSectionStringType"/> or creates one
+        /// if the type does not exist yet for the given <paramref name="text"/>.
+        /// If the text cannot be encoded, returns <see langword="null"/>.
+        /// </summary>
+        internal static Cci.IFieldReference? TryGetOrCreateFieldForStringValue<TArg>(
+            string text,
+            TArg arg,
+            Func<TArg, PrivateImplementationDetails> factory,
+            DiagnosticBag diagnostics)
+        {
+            if (!text.TryGetUtf8ByteRepresentation(out byte[]? data, out _))
+            {
+                return null;
+            }
+
+            var @this = factory(arg);
+            return @this._dataSectionStringLiteralTypes.GetOrAdd(text, static (key, arg) =>
+            {
+                var (@this, data, diagnostics) = arg;
+
+                string name = "<S>" + DataToHexViaXxHash128(data);
+
+                MappedField dataField = @this.CreateDataField(data, alignment: 1);
+
+                Cci.IMethodDefinition bytesToStringHelper = @this.GetOrSynthesizeBytesToStringHelper(@this.ModuleBuilder.CommonCompilation, diagnostics);
```

Comment:
>@this.ModuleBuilder.CommonCompilatio [](http://example.com/codeflow?start=101&length=36)

It looks like we are calling an instance method and explicitly passing an information that is directly available from the instance.